server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java

```diff
@@ -0,0 +1,309 @@
+
+                            if (logger.isDebugEnabled()) {
+                                key.pollEvents().stream().forEach(e -> logger.debug("{}:{}", e.kind().toString(), e.context().toString()));
+                            }
+
+                            key.pollEvents();
+                            key.reset();
+
+                            enableSettingsWatcher(settingsDir);
+
+                            if (watchedFileChanged(path)) {
+                                processFileSettings(path, false);
+                            }
+                        } catch (Exception e) {
+                            logger.warn("unable to watch or read operator settings file", e);
+                        }
+                    } else {
+                        key.pollEvents();
+                        key.reset();
+                    }
+                }
+            } catch (Exception e) {
+                if (logger.isDebugEnabled()) {
+                    logger.debug("encountered exception watching", e);
+                }
+                logger.info("shutting down watcher thread");
+            } finally {
+                watcherThreadLatch.countDown();
+            }
+        }, "elasticsearch[file-settings-watcher]").start();
+    }
+
+    synchronized void stopWatcher() {
+        logger.debug("stopping watcher ...");
+        if (watching()) {
+            try {
+                watchService.close();
+                if (watcherThreadLatch != null) {
+                    watcherThreadLatch.await();
+                }
+            } catch (IOException | InterruptedException e) {
+                logger.info("encountered exception while closing watch service", e);
+            } finally {
+                watchService = null;
+                logger.info("watcher service stopped");
+            }
+        } else {
+            logger.debug("file settings service already stopped");
+        }
+    }
+
+    private void enableSettingsWatcher(Path settingsDir) throws IOException {
+        settingsDir.register(
+            watchService,
+            StandardWatchEventKinds.ENTRY_MODIFY,
+            StandardWatchEventKinds.ENTRY_CREATE,
+            StandardWatchEventKinds.ENTRY_DELETE
+        );
+    }
+
+    void processFileSettings(Path path, boolean onStartup) {
+        logger.info("processing path [{}] for [{}]", path, NAMESPACE);
+        try (
+            XContentParser parser = XContentType.JSON.xContent().createParser(XContentParserConfiguration.EMPTY, Files.newInputStream(path))
+        ) {
+            stateService.process(NAMESPACE, parser, (e) -> {
+                if (e != null) {
+                    // If we encountered an exception trying to apply the operator state at
+                    // startup time, we throw an error to force Elasticsearch to exit.
+                    if (onStartup) {
+                        throw new OperatorConfigurationError("Error applying operator settings", e);
```

Comment:
It looks like this gets propagated because the exception consumer is not guarded. That's fine, but then we catch the exception down below, so it seems like we would not fail startup as intended? Am I missing something?

Reply:
OperatorConfigurationError is of type Error, so it's never caught. I wasn't sure if there's a nicer way to cause the exception to skip all the exception handling we do and terminate the node.